text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,484 @@
+---
+feature name: CloudFront redesign
+start date: 2020-06-15
+rfc pr:
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/171
+---
+
+# Summary
+
+(Draft) Proposal to redesign the @aws-cdk/aws-cloudfront module.
+
+The current module does not adhere to the best practice naming conventions or ease-of-use patterns
+that are present in the other CDK modules. A redesign of the API will allow for friendly, easier
+access to common patterns and usages.
+
+This RFC does not attempt to lay out the entire API; rather, it focuses on a complete re-write of
+the module README with a focus on the most common use cases and how they work with the new design.
+More detailed designs and incremental API improvements will be tracked as part of GitHub Project Board
+once the RFC is approved.
+
+---
+
+# README
+
+Amazon CloudFront is a web service that speeds up distribution of your static and dynamic web content, such as .html, .css, .js, and image files, to
+your users. CloudFront delivers your content through a worldwide network of data centers called edge locations. When a user requests content that
+you're serving with CloudFront, the user is routed to the edge location that provides the lowest latency, so that content is delivered with the best
+possible performance.
+
+## Creating a distribution
+
+CloudFront distributions deliver your content from one or more origins; an origin is the location where you store the original version of your
+content. Origins can be created from S3 buckets or a custom origin (HTTP server).
+
+An S3 bucket can be added as an origin. If the bucket is configured as a website endpoint, the distribution can use S3 redirects and S3 custom error
+documents.
+
+```ts
+import * as cloudfront from '@aws-cdk/aws-cloudfront';
+
+// Creates a distribution for a S3 bucket.
+const myBucket = new s3.Bucket(...);
+new cloudfront.Distribution(this, 'myDist', {
+  origin: cloudfront.Origin.fromBucket(myBucket),
```

Comment:
Taking that suggestion of syntactic sugar a bit further, would changing `Origin.fromBucket` to `Origin.forBucket` reduce some of the dissonance here? Also, I like top-level sugar (`Distribution.forBucket`); would you suggest that for just the most common (e.g., forBucket / forWebsiteBucket) or would we end up with one `Distribution.for*` for every `Origin.for*`?

Reply:
> would changing Origin.fromBucket to Origin.forBucket reduce some of the dissonance here?

I considered that, but `Origin.forBucket` doesn't really feel right. `Distribution.forBucket` makes sense because the distribution is serving the bucket in a way. I prefer the `from` terminology on `Origin`, the are actually semantically related :)

Lets see where we end with on the `from` usage and we can change later on. 

> would you suggest that for just the most common (e.g., forBucket / forWebsiteBucket) or would we end up with one Distribution.for* for every Origin.for*?

I think just for the common ones, those that we need minimum configuration for. I'm ok with it also being just for bucket. Definitely not a must to do it for all.